### ISSUE TITLE
PDB are being ignored when the node pools have to be recreated.

### ISSUE THREAD

**Issue by justinb-shipt** [terraform-google-modules/terraform-google-kubernetes-engine#1456]
### TL;DR

When node pools have to be recreated because of newer features, PDB for critical services are ignored.

### Expected behavior

I would hope that if node pools are recreated, it would use the eviction API so that PDB are respected.

### Observed behavior

We configure our GKE clusters with the Terraform GKE module. We were on version v15.0.0 and have been upgrading to v21.2.0. Between these changes, there have been additional fields (keepers on random_id resource) added to help manage/configure node pools. Because of these additional fields, TF decided it had to rebuild all the node pools to meet the new spec. 

```
  # module.gke.random_id.name["workers-01"] must be replaced
+/- resource "random_id" "name" {
      ~ b64_std     = "workers-01-4Mc=" -> (known after apply)
      ~ b64_url     = "workers-01-4Mc" -> (known after apply)
      ~ dec         = "workers-01-57543" -> (known after apply)
      ~ hex         = "workers-01-e0c7" -> (known after apply)
      ~ id          = "4Mc" -> (known after apply)
      ~ keepers     = { # forces replacement
          + "enable_gcfs"        = ""
          + "enable_secure_boot" = ""
          + "min_cpu_platform"   = ""
          + "spot"               = ""
            # (13 unchanged elements hidden)
        }
        # (2 unchanged attributes hidden)
    }
```

That’s all well and good. In this example, `workers-01` is going to get replaced with a new `workers-01`. We first got a clue of a potential issue when it came to our Kafka and Vault clusters. Both have PDBs saying that minUnavailable is 1 and allowedDisruptions is 1. However, during the apply, we had outages with Kafka and Vault. I watched the upgrade on the Vault cluster more carefully and saw that Consul was moved between node pools and that all three Pods were there but not ready. All former Pods were gone. So it appears that when node pools are recreated it's not using the eviction API(?) so PDB are ignored. So it's treating the node pools as 100% new and not replacements? Then the old node pools are destroyed and chaos.

With other Pods w/o PDBs, they will enter a crash loop state because they were also all evicted at once and waiting on Nodes. That creates a condition were we have to restart those Pods. This might be able to be solved by updating our minimums in our node pools to match the current count of the cluster, but it would be much better if the Pods were moved more gracefully.

### Terraform Configuration

```hcl
# google_client_config and kubernetes provider must be explicitly specified like the following.
data "google_client_config" "default" {}

provider "kubernetes" {
  host                   = "https://${module.gke.endpoint}"
  token                  = data.google_client_config.default.access_token
  cluster_ca_certificate = base64decode(module.gke.ca_certificate)
}

module "gke" {
  source                     = "terraform-google-modules/kubernetes-engine/google"
  version                    = "v21.2.0"

  project_id                 = "<PROJECT ID>"
  name                       = "gke-test-1"
  region                     = "us-central1"
  zones                      = ["us-central1-a", "us-central1-b", "us-central1-f"]
  network                    = "vpc-01"
  subnetwork                 = "us-central1-01"
  ip_range_pods              = "us-central1-01-gke-01-pods"
  ip_range_services          = "us-central1-01-gke-01-services"
  http_load_balancing        = false
  network_policy             = false
  horizontal_pod_autoscaling = true
  filestore_csi_driver       = false

  node_pools = [
    {
      name                      = "default-node-pool"
      machine_type              = "e2-medium"
      node_locations            = "us-central1-b,us-central1-c"
      min_count                 = 1
      max_count                 = 100
      local_ssd_count           = 0
      spot                      = false
      disk_size_gb              = 100
      disk_type                 = "pd-standard"
      image_type                = "COS_CONTAINERD"
      enable_gcfs               = false
      enable_gvnic              = false
      auto_repair               = true
      auto_upgrade              = true
      service_account           = "project-service-account@<PROJECT ID>.iam.gserviceaccount.com"
      preemptible               = false
      initial_node_count        = 80
    },
  ]

  node_pools_oauth_scopes = {
    all = [
      "https://www.googleapis.com/auth/logging.write",
      "https://www.googleapis.com/auth/monitoring",
    ]
  }

  node_pools_labels = {
    all = {}

    default-node-pool = {
      default-node-pool = true
    }
  }

  node_pools_metadata = {
    all = {}

    default-node-pool = {
      node-pool-metadata-custom-value = "my-node-pool"
    }
  }

  node_pools_taints = {
    all = []

    default-node-pool = [
      {
        key    = "default-node-pool"
        value  = true
        effect = "PREFER_NO_SCHEDULE"
      },
    ]
  }

  node_pools_tags = {
    all = []

    default-node-pool = [
      "default-node-pool",
    ]
  }
}
```

### Terraform Version

```sh
# We actually run in Terraform Cloud v1.3.3.
❯ terraform version
Terraform v1.3.0
on darwin_amd64
+ provider registry.terraform.io/gavinbunney/kubectl v1.11.3
+ provider registry.terraform.io/hashicorp/google v3.80.0
+ provider registry.terraform.io/hashicorp/helm v2.2.0
+ provider registry.terraform.io/hashicorp/kubernetes v2.4.1
+ provider registry.terraform.io/hashicorp/time v0.7.2
+ provider registry.terraform.io/hashicorp/tls v3.1.0
+ provider registry.terraform.io/hashicorp/vault v2.22.1
```

**Comment by wintergren** [comment]
We're in the same situation and had expected PDB's to be respected but sadly not so experiencing some downtime now during node pool upgrades. 

Would seem the module needs to use the eviction api as mentioned by @justinb-shipt https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#eviction-api

**Comment by justinb-shipt** [comment]
Are there any different setting that could be tried on the node pools themselves?

**Comment by wintergren** [comment]
Updated to GKE v1.22 yesterday and with that upgrade we did not experience this issue, so could be that it only affects v1.21.x and earlier versions of K8s?

**Comment by justinb-shipt** [comment]
Our Kafka cluster was already on GKE v1.22 and we still saw this issue.

**Comment by justinb-shipt** [comment]
@wintergren, could you do me a favor and log into one of your nodes and run `systemd-inhibit --list`? I'm trying to see if the potential difference is that you have [Graceful Node Shutdown](https://kubernetes.io/blog/2021/04/21/graceful-node-shutdown-beta/) enabled and I don't.

**Comment by justinb-shipt** [comment]
I dug down into the code and it looks like this might be more on the Google Provider than this module. Read the [comments](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/modules/beta-private-cluster-update-variant/cluster.tf#L467-L470) for the `random_id` resource and it points you to the Google Provider. If you specifically look at [schemaNodeConfig](https://github.com/hashicorp/terraform-provider-google/blob/main/google/node_config.go#L49) you'll see the list of keepers and the setting `ForceNew`.

**Comment by florianmutter** [comment]
Might be a duplicate of #794

**Comment by nemethloci** [comment]
> Updated to GKE v1.22 yesterday and with that upgrade we did not experience this issue, so could be that it only affects v1.21.x and earlier versions of K8s?

I've just upgraded multiple clusters from 1.23 to 1.24 and I still still faced this issue.
